src/1-raw-data/mailinglists/main.go

```
@@ -0,0 +1,145 @@
+}
+
+func (gcs *storageConnection) connectGCSClient() error {
+	if client, err := storage.NewClient(gcs.ctx); err != nil {
+		return fmt.Errorf("Failed to create client: %v", err)
+	} else {
+		gcs.client = client
+		return nil
+	}
+}
+
+// Creates storage bucket if it doesn't exist.
+func (gcs *storageConnection) createGCSBucket() error {
+	// Setup client bucket to work from
+	gcs.bucket = gcs.client.Bucket(*bucketName)
+
+	buckets := gcs.client.Buckets(gcs.ctx, *projectID)
+	for {
+		attrs, err := buckets.Next()
+		// Assume that if Iterator end then not found and need to create bucket
+		if err == iterator.Done {
+			// Create bucket if it doesn't exist - https://cloud.google.com/storage/docs/reference/libraries
+			if err := gcs.bucket.Create(gcs.ctx, *projectID, &storage.BucketAttrs{
+				Location: "US",
+			}); err != nil {
+				// TODO - add random number to append to bucket name to resolve
+				return fmt.Errorf("Failed to create bucket: %v", err)
+			}
+			log.Printf("Bucket %v created.\n", *bucketName)
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("Issues setting up Bucket(%q).Objects(): %v. Double check project id.", attrs.Name, err)
+		}
+		if attrs.Name == *bucketName {
+			//getLatestFile() // TODO set this up to check and compare what is in the bucket vs what isn't
+			log.Printf("Bucket %v exists.\n", bucketName)
+			return nil
+		}
+	}
+}
+
+// Store files in storage.
+func (gcs *storageConnection) storeGCS(fileName string, url string) {
+	// Get HTTP response
+	response, _ := http.Get(url)
+	defer response.Body.Close()
+
+	if response.StatusCode == http.StatusOK {
+		obj := gcs.bucket.Object(fileName)
+
+		// w implements io.Writer.
+		w := obj.NewWriter(gcs.ctx)
+
+		// Copy file into storage
+		_, err := io.Copy(w, response.Body)
+		if err != nil {
+			log.Printf("Failed to copy %v to bucket: %v", fileName, err)
+		}
+		response.Body.Close()
+
+		if err := w.Close(); err != nil {
+			log.Fatalf("Failed to close: %v", err)
```

instead of failing hard here i'd let this function return an error